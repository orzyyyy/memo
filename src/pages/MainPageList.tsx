import React from 'react';
import { format } from 'date-fns';
import { MappingProps } from '../../server/controller/DocumentController';
import { DocumentCategoryProps } from '../../server/utils/document';

const MainPageList = ({
  onDelete,
  onEdit,
  onHide,
  siderSelectedKey,
  dataSource,
  onListItemClick,
  isLocal,
}: {
  siderSelectedKey: string;
  onDelete?: (dataItem: MappingProps) => void;
  onHide?: (dataItem: MappingProps) => void;
  dataSource: MappingProps[];
  onEdit: (dataItem?: MappingProps, visible?: boolean, pageInfo?: { x: number; y: number }) => void;
  onListItemClick: ({ category, id }: { category: DocumentCategoryProps; id: string }) => void;
  isLocal: boolean;
}) => {
  const data = siderSelectedKey === 'all' ? dataSource : dataSource.filter(item => item.subType === siderSelectedKey);

  return (
    <ul>
      {data.map(item => {
        const buttonGroup = (
          <div style={{ float: 'right', marginRight: 8 }}>
            <a
              onClick={(e: React.MouseEvent) => {
                e.stopPropagation();
                onEdit(item, true);
              }}
              style={{ marginRight: 16 }}
            >
              修改
            </a>
            <a
              onClick={(e: React.MouseEvent) => {
                e.stopPropagation();
                onHide && onHide(item);
              }}
              style={{ marginRight: 16 }}
            >
              隐藏
            </a>
            <a
              onClick={(e: React.MouseEvent) => {
                e.stopPropagation();
                onDelete && onDelete(item);
              }}
              style={{ marginRight: 16 }}
            >
              删除
            </a>
          </div>
        );

        return (
          <li
            key={`list-item-${item.id}`}
            className="list-item"
            onClick={() => onListItemClick({ category: item.category, id: item.id })}
          >
            {item.category === 'mapping' && <div style={{ background: '#108ee9' }} className="icon-apartment" />}
            {item.category === 'markdown' && <div style={{ background: '#87d068' }} className="icon-file-markdown" />}
            {item.type + ' - ' + item.subType + ' - ' + item.title}
            <div style={{ float: 'right', marginRight: 8 }}>
              {`${format(new Date(item.createTime || ''), 'yyyy-MM-dd')} / ${format(
                new Date(item.modifyTime || ''),
                'yyyy-MM-dd',
              )}`}
            </div>
            {isLocal && buttonGroup}
          </li>
        );
      })}
    </ul>
  );
};

export default MainPageList;
