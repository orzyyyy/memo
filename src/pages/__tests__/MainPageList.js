import React from 'react';
import { mount } from 'enzyme';
import MainPageList from '../MainPageList';

const dataSource = [
  {
    id: '32e89f38116080d6b7608cf2137f3c96',
    title: '音程表',
    url: './assets/mapping/32e89f38116080d6b7608cf2137f3c96.json',
    createTime: 1559522075108,
    modifyTime: 1559523700429,
    type: '音乐',
    subType: '乐理',
    category: 'markdown',
  },
  {
    id: '7a73d21c13f1eb55053d4af66e0d057b',
    title: 'windows 下安装 rust',
    url: './assets/mapping/7a73d21c13f1eb55053d4af66e0d057b.json',
    createTime: 1559617097664,
    modifyTime: 1559633639037,
    type: 'Rust',
    subType: '安装',
    category: 'mapping',
  },
];

describe('MainPageList', () => {
  it('render correctly', () => {
    const onDelete = jest.fn();
    const onEdit = jest.fn();
    const props = {
      dataSource,
      onDelete,
      onEdit,
    };
    const state = {
      siderSelectedKey: 'all',
    };
    const handleListItemClick = jest.fn();
    const event = {
      handleListItemClick,
    };
    const wrapper = mount(
      <MainPageList props={props} state={state} event={event} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('events works correctly', () => {
    const onDelete = jest.fn();
    const onEdit = jest.fn();
    const props = {
      dataSource,
      onDelete,
      onEdit,
    };
    const state = {
      siderSelectedKey: 'all',
    };
    const handleListItemClick = jest.fn();
    const event = {
      handleListItemClick,
    };
    const wrapper = mount(
      <MainPageList props={props} state={state} event={event} />,
    );
    wrapper
      .find('Item')
      .first()
      .simulate('contextmenu');
    wrapper
      .find('.ant-dropdown-menu-item')
      .first()
      .props()
      .onClick();
    expect(onEdit).toHaveBeenCalled();
    wrapper
      .find('.ant-dropdown-menu-item')
      .last()
      .props()
      .onClick();
    expect(onDelete).toHaveBeenCalled();
    wrapper
      .find('Item')
      .first()
      .simulate('click');
    expect(handleListItemClick).toHaveBeenCalled();
  });
});
