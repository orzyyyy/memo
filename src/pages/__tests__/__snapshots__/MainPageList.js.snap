// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPageList render correctly 1`] = `
<List
  bordered={false}
  dataSource={
    Array [
      Object {
        "category": "markdown",
        "createTime": 1559522075108,
        "id": "32e89f38116080d6b7608cf2137f3c96",
        "modifyTime": 1559523700429,
        "subType": "乐理",
        "title": "音程表",
        "type": "音乐",
        "url": "./assets/mapping/32e89f38116080d6b7608cf2137f3c96.json",
      },
      Object {
        "category": "markdown",
        "createTime": 1559617097664,
        "id": "7a73d21c13f1eb55053d4af66e0d057b",
        "modifyTime": 1559633639037,
        "subType": "安装",
        "title": "windows 下安装 rust",
        "type": "Rust",
        "url": "./assets/mapping/7a73d21c13f1eb55053d4af66e0d057b.json",
      },
    ]
  }
  loading={false}
  pagination={false}
  renderItem={[Function]}
  split={true}
/>
`;
